positron/modules/atom_common_id_weak_map.js

```diff
@@ -0,0 +1,24 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+"use strict";
+
+const { classes: Cc, interfaces: Ci, results: Cr, utils: Cu } = Components;
+
+// Described In Electron's id_weak_map.h as:
+// "Like ES6's WeakMap, but the key is Integer and the value is Weak Pointer."
+
+function IDWeakMap() {
+  const map = new Map();
+  this.set = function(key, value) {
+    map.set(key, Cu.getWeakReference(value));
+  };
+  this.get = function(key) {
+    return map.get(key).get();
+  }
+  this.has = map.has.bind(map);
+  this.clear = map.clear.bind(map);
+}
```

Comment:
> On 20 May 2016, at 2:51 AM, Myk Melez <notifications@github.com> wrote:
> 
> In positron/modules/atom_common_id_weak_map.js:
> 
> > +const { classes: Cc, interfaces: Ci, results: Cr, utils: Cu } = Components;
> > +
> > +// Described In Electron's id_weak_map.h as:
> > +// "Like ES6's WeakMap, but the key is Integer and the value is Weak Pointer."
> > +
> > +function IDWeakMap() {
> > +  const map = new Map();
> > +  this.set = function(key, value) {
> > +    map.set(key, Cu.getWeakReference(value));
> > +  };
> > +  this.get = function(key) {
> > +    return map.get(key).get();
> > +  }
> > +  this.has = map.has.bind(map);
> > +  this.clear = map.clear.bind(map);
> > +}
> Do you know if it matters that these properties are on the instance rather than the prototype?
> 

It might. We should check if anyone has used the Electron version by calling the methods by going through the prototype. 

> If we want them on the prototype, we might need something like this:
> 
> I don't think it matters. In theory, the prototype-based implementation is more performant, but I don't know if it's actually more performant in practice. And I don't think this is performance-sensitive code (i.e. that the number of IDWeakMap instances is high enough for a difference in construction perf to matter). However, I'm happy switch to it anyway, if you prefer it for another reason.
> 

As above. Not really a perf concern, just a interop one. However, we could certainly see how it goes with the current implementation. 
> —
> You are receiving this because you were assigned.
> Reply to this email directly or view it on GitHub
>